lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,881 @@
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (sender)
+fn calculate_act_message(local_private_ephemeral_key: &SecretKey, local_public_ephemeral_key: &PublicKey, remote_public_key: &PublicKey, chaining_key: ChainingKey, hash: Sha256, act_out: &mut [u8]) -> (Sha256, SymmetricKey, SymmetricKey) {
+	// 1. e = generateKey() (passed in)
+	// 2. h = SHA-256(h || e.pub.serializeCompressed())
+	let serialized_local_public_key = local_public_ephemeral_key.serialize();
+	let hash = concat_then_sha256!(hash, serialized_local_public_key);
+
+	// 3. ACT1: es = ECDH(e.priv, rs)
+	// 3. ACT2: es = ECDH(e.priv, re)
+	let ecdh = ecdh(local_private_ephemeral_key, &remote_public_key);
+
+	// 4. ACT1: ck, temp_k1 = HKDF(ck, es)
+	// 4. ACT2: ck, temp_k2 = HKDF(ck, ee)
+	let (chaining_key, temporary_key) = hkdf::derive(&chaining_key, &ecdh);
+
+	// 5. ACT1: c = encryptWithAD(temp_k1, 0, h, zero)
+	// 5. ACT2: c = encryptWithAD(temp_k2, 0, h, zero)
+	chacha::encrypt(&temporary_key, 0, &hash, &[0; 0], &mut act_out[34..]);
+
+	// 6. h = SHA-256(h || c)
+	let hash = concat_then_sha256!(hash, &act_out[34..]);
+
+	// Send m = 0 || e.pub.serializeCompressed() || c
+	act_out[1..34].copy_from_slice(&serialized_local_public_key);
+
+	(hash, chaining_key, temporary_key)
 }
+
+// Due to the very high similarity of acts 1 and 2, this method is used to process both
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (receiver)
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (receiver)
+fn process_act_message(act_bytes: &[u8], local_private_key: &SecretKey, chaining_key: ChainingKey, hash: Sha256) -> Result<(PublicKey, Sha256, SymmetricKey, SymmetricKey), String> {
+	// 1. Read exactly 50 bytes from the network buffer
+	// 2.Parse the read message (m) into v, re, and c
+	let version = act_bytes[0];
+	let ephemeral_public_key_bytes = &act_bytes[1..34];
+	let chacha_tag = &act_bytes[34..];
+
+	let ephemeral_public_key = if let Ok(public_key) = PublicKey::from_slice(&ephemeral_public_key_bytes) {
+		public_key
+	} else {
+		return Err("invalid remote ephemeral public key".to_string());
+	};
+
+	// 3. If v is an unrecognized handshake version, then the responder MUST abort the connection attempt
+	if version != 0 {
+		// this should not crash the process, hence no panic
+		return Err("unexpected version".to_string());
+	}
+
+	// 4. h = SHA-256(h || re.serializeCompressed())
+	let hash = concat_then_sha256!(hash, ephemeral_public_key_bytes);
+
+	// 5. Act1: es = ECDH(s.priv, re)
+	// 5. Act2: ee = ECDH(e.priv, ee)
+	let ecdh = ecdh(local_private_key, &ephemeral_public_key);
+
+	// 6. Act1: ck, temp_k1 = HKDF(ck, es)
+	// 6. Act2: ck, temp_k2 = HKDF(ck, ee)
+	let (chaining_key, temporary_key) = hkdf::derive(&chaining_key, &ecdh);
+
+	// 7. p = decryptWithAD(temp_k1, 0, h, c)
```

Comment:
Same as step above, it should be duplicated with a `7. p = decryptWithAD(temp_k2, 0, h, c)`

Reply:
Resolved in `review: Non-controversial renames, nits, small fixups`